Bluetooth: Mesh: Support for comp data page 2

Andrewpini opened the pull request:
- Adds support for composition data page 2 &129.
- Adds testing of composition data page 2 and 130 in the
mesh_prov_pst_pb_remote_ncrp bsim test.

Andrewpini commented:
This draft is based on https://github.com/zephyrproject-rtos/zephyr/pull/61794 which is pending review. There is no need for a thorough review here. I want a discussion on the conceptual spilt between upstream and NCS, and how it will affect each of them. When we agree on the concept I will create a new PR to zephyr main.

FYI: This passes Bsim on my machine.

omkar3141 commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2122,6 +2150,24 @@ int bt_mesh_comp_store(void)
+		err = bt_mesh_comp_data_get_page_2(&buf);
+
+		if (err) {
+			LOG_ERR("Failed to read composition data P2: %d", err);
```
When you actually create PR, it may be worth to combine all these error messages. There is quite a bit of repeated code now. If you want to keep the structure, may be rewording these strings will optimize flash a bit. E.g. "read err cdp2: %d".

omkar3141 reviewed (approved):
Minor comment. The overall approach looks fine to me.

PavelVPV commented on include/zephyr/bluetooth/mesh/cfg_cli.h:
```diff
@@ -1743,6 +1743,22 @@ struct bt_mesh_comp_p1_ext_item {
 	};
 };
 
+struct bt_mesh_comp_p2_elem {
```
Missing description for the struct and some other fields.

PavelVPV commented on include/zephyr/bluetooth/mesh/cfg_cli.h:
```diff
@@ -1743,6 +1743,22 @@ struct bt_mesh_comp_p1_ext_item {
+struct bt_mesh_comp_p2_elem {
+	uint16_t id;
+	struct {
+		/** Major version*/
+		uint8_t x;
+		/** Minor version*/
+		uint8_t y;
+		/** Z version*/
```
```suggestion
		/** Major version */
		uint8_t x;
		/** Minor version */
		uint8_t y;
		/** Z version */
```

PavelVPV commented on include/zephyr/bluetooth/mesh/access.h:
```diff
@@ -1118,6 +1118,24 @@ bool bt_mesh_model_is_extended(struct bt_mesh_model *model);
+ *
+ *  @param cb Callback structure.
+ */
+void bt_mesh_comp_data_page_2_cb_reg(bt_mesh_comp2_cb_t cb);
```
Do you see any usability with the callback for this? I feel this a bit unusual to have a callback for the data that is to be constant during the device run time. 

What if we add a function like `bt_mesh_comp_register` but external that would register pointer to the cdp2? Then we avoid calling callback every time the cdp2 is requested. The cdp2 itself should be registered on the user side like cdp0. It can be (or even should be) declared as `const` because it won't change at runtime (at least the API should only take const data).

I'm also thinking that it should be registered not as a `net_buf_simple` but as a decoded structure so that it is easier for user to create and maintain it and that it allows users to avoid packing it manually or calling additional API to encode it into `net_buf_simple`. Unless we can use some preprocessor magic to pack it at compile time.

Andrewpini replied:
I like this 👍 I will do another iteration on implementation.

omkar3141 replied:
It is very unlikely that preprocessor magic will work. Registering pointer has a small risk that pointer needs to be filled in appropriately or valid, may be this is not a concern since NCS implementation that calls this API will make sure of that.

PavelVPV replied:
I think it is the same risk level as calling any of our API with NULL pointer, or calling API from a preemptable thread. This is how we instantiate CDP0 and should work well for CDP2. Perhaps this is the solution to avoid storing CDP1 in RAM as we do now: https://github.com/zephyrproject-rtos/zephyr/blob/a15e7d2326d9e8896401be479fae56f30dde71bb/subsys/bluetooth/mesh/access.c#L82?

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -2122,6 +2150,24 @@ int bt_mesh_comp_store(void)
 		}
 	}
+
+	if (IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_2)) {
```
Now, with CDP2 it is clear that we need to optimize this code a bit. What about this? https://github.com/zephyrproject-rtos/zephyr/commit/f1da818d965f681a88308d34c802b8908ca1f3f7
I didn't test it, only compiled. This can probably be added in CDP129 PR.

PavelVPV replied:
I was able to cut ~700 bytes with this commit.

PavelVPV replied:
Sorry, that commit requires this patch:
```
diff --git a/subsys/bluetooth/mesh/cfg_srv.c b/subsys/bluetooth/mesh/cfg_srv.c
index 5258fed8db..9a1f6e6131 100644
--- a/subsys/bluetooth/mesh/cfg_srv.c
+++ b/subsys/bluetooth/mesh/cfg_srv.c
@@ -87,9 +87,9 @@ static int dev_comp_data_get(struct bt_mesh_model *model,
        net_buf_simple_add_u8(&sdu, page);
 
        if (atomic_test_bit(bt_mesh.flags, BT_MESH_COMP_DIRTY) && page < 128) {
-               buf->size -= BT_MESH_MIC_SHORT;
-               err = bt_mesh_comp_read(buf, page);
-               buf->size += BT_MESH_MIC_SHORT;
+               sdu.size -= BT_MESH_MIC_SHORT;
+               err = bt_mesh_comp_read(&sdu, page);
+               sdu.size += BT_MESH_MIC_SHORT;
```

PavelVPV replied:
To implement this: https://github.com/zephyrproject-rtos/zephyr/pull/61794/files#r1308733333, you only need this:
```
diff --git a/subsys/bluetooth/mesh/access.c b/subsys/bluetooth/mesh/access.c
index 5045a0992c..059c4aeca4 100644
--- a/subsys/bluetooth/mesh/access.c
+++ b/subsys/bluetooth/mesh/access.c
@@ -109,12 +109,12 @@ static const struct {
        uint8_t *path;
        uint8_t page;
 } comp_data_pages[] = {
-       { "bt/mesh/cmp", 0, },
+       { "bt/mesh/cmp/0", 0, },
 #if IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_1)
-       { "bt/mesh/cmp1", 1, },
+       { "bt/mesh/cmp/1", 1, },
 #endif
 #if IS_ENABLED(CONFIG_BT_MESH_COMP_PAGE_2)
-       { "bt/mesh/cmp2", 2, },
+       { "bt/mesh/cmp/2", 2, },
 #endif
 };
```
And restore old `comp_set` function.

Andrewpini replied:
Yes I will add it in the other PR. I will rebase this PR/draft when CDP129 PR is merged.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -1407,6 +1438,24 @@ static void test_provisioner_pb_remote_client_ncrp(void)
 		FAIL("Wrong composition data page 129");
 	}
 
+	/* Check that Composition Data Page 130 still exists and is now equal to Page 2. */
```
These checks are repeated for each page pair. Could this be wrapped into a function?

Andrewpini replied:
Yes, I will add it in the other PR 👍

Andrewpini committed:
Bluetooth: Mesh: Support for comp data page 2

Adds support for composition data page 2 &129.

In this implementation the responsibillity for filling the page 2
buffer is left to the application through the new comp page 2 cb API.
Only the application can know/decide if the device is NLC compliant,
and must thus be given the responsibillity for cheking the NLC profile
requirements, defined in the NLC specs, and filling the response buffer
for comp data page 2.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
tests: Bluetooth: Mesh: Add comp p2 test to bsim

Adds testing of composition data page 2 and 130 in the
mesh_prov_pst_pb_remote_ncrp bsim test.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: Shell support for comp data page2

Adds shell support for composition data page 2 and 130

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -1325,37 +1354,63 @@ static void test_provisioner_pb_remote_client_ncrp_provision(void)
+			FAIL("Composition data is not equal");
+		}
+	} else {
+		if (comp1->len == comp2->len) {
+			if (!memcmp(comp1->data, comp2->data, comp1->len)) {
+				FAIL("Composition data is equal");
+			}
+		}
```
```suggestion
		ASSERT_TRUE(comp1->len != comp2->len);
		if (!memcmp(comp1->data, comp2->data, comp1->len)) {
			FAIL("Composition data is equal");
		}
```

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -181,6 +181,31 @@ static const struct bt_mesh_comp rpr_srv_comp_unresponsive = {
+static uint8_t elem_offset2[3] = {4, 5, 6};
+static uint8_t additional_data[2] = {100, 200};
+
+static struct bt_mesh_comp2_record comp_rec[2] = {
+	{.id = 1,
+	 .version.x = 2,
+	 .version.y = 3,
+	 .version.z = 4,
+	 .elem_offset_cnt = sizeof(elem_offset1),
+	 .elem_offset = elem_offset1,
+	 .additional_data_len = 0},
+	{.id = 10,
+	 .version.x = 20,
+	 .version.y = 30,
+	 .version.z = 40,
+	 .elem_offset_cnt = sizeof(elem_offset2),
+	 .elem_offset = elem_offset2,
+	 .additional_data_len = sizeof(additional_data),
+	 .data = additional_data},
+};
```
```suggestion
static struct bt_mesh_comp2_record comp_rec[2] = {
	{
		.id = 1,
		.version.x = 2,
		.version.y = 3,
		.version.z = 4,
		.elem_offset_cnt = sizeof(elem_offset1),
		.elem_offset = elem_offset1,
		.additional_data_len = 0
	},
	{
		.id = 10,
		.version.x = 20,
		.version.y = 30,
		.version.z = 40,
		.elem_offset_cnt = sizeof(elem_offset2),
		.elem_offset = elem_offset2,
		.additional_data_len = sizeof(additional_data),
		.data = additional_data
	},
};
```

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -181,6 +181,31 @@ static const struct bt_mesh_comp rpr_srv_comp_unresponsive = {
+	 .data = additional_data},
+};
+
+static struct bt_mesh_comp2 comp_p2_1 = {.record_cnt = 1, .rec = comp_rec};
+static struct bt_mesh_comp2 comp_p2_2 = {.record_cnt = 2, .rec = comp_rec};
```
```suggestion
static const struct bt_mesh_comp2 comp_p2_1 = {.record_cnt = 1, .rec = comp_rec};
static const struct bt_mesh_comp2 comp_p2_2 = {.record_cnt = 2, .rec = comp_rec};
```

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -181,6 +181,31 @@ static const struct bt_mesh_comp rpr_srv_comp_unresponsive = {
+static uint8_t elem_offset2[3] = {4, 5, 6};
+static uint8_t additional_data[2] = {100, 200};
+
+static struct bt_mesh_comp2_record comp_rec[2] = {
```
```suggestion
static const struct bt_mesh_comp2_record comp_rec[2] = {
```

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -181,6 +181,31 @@ static const struct bt_mesh_comp rpr_srv_comp_unresponsive = {
 	.elem_count = 1,
 };
 
+static uint8_t elem_offset1[2] = {1, 2};
+static uint8_t elem_offset2[3] = {4, 5, 6};
+static uint8_t additional_data[2] = {100, 200};
```
```suggestion
static const uint8_t elem_offset1[2] = {1, 2};
static const uint8_t elem_offset2[3] = {4, 5, 6};
static const uint8_t additional_data[2] = {100, 200};
```

Andrewpini commented:
https://github.com/zephyrproject-rtos/zephyr/pull/61794 is now merged, so I will close this draft an create a new PR to zephyr main.

Andrewpini commented:
New PR: https://github.com/zephyrproject-rtos/zephyr/pull/62110
